src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,272 @@
+                    var formattingService = document?.GetLanguageService<IEditorFormattingService>();
+
+                    if (formattingService != null)
+                    {
+                        var changes = formattingService.GetFormattingChangesAsync(
+                            document, spanToFormat.Span.ToTextSpan(), CancellationToken.None).WaitAndGetResult(CancellationToken.None);
+                        document.Project.Solution.Workspace.ApplyTextChanges(document.Id, changes, CancellationToken.None);
+                    }
+                }
+            }
+
+            if (includesCommitCharacter)
+            {
+                return AsyncCompletionData.CommitBehavior.SuppressFurtherTypeCharCommandHandlers;
+            }
+
+            if (commitCharacter == '\n' && SendEnterThroughToEditor(rules, roslynItem, filterText))
+            {
+                return AsyncCompletionData.CommitBehavior.RaiseFurtherReturnKeyAndTabKeyCommandHandlers;
+            }
+
+            return AsyncCompletionData.CommitBehavior.None;
+        }
+
+        internal static bool IsCommitCharacter(CompletionRules completionRules, CompletionItem item, char ch, string textTypedSoFar)
+        {
+            // First see if the item has any specifc commit rules it wants followed.
+            foreach (var rule in item.Rules.CommitCharacterRules)
+            {
+                switch (rule.Kind)
+                {
+                    case CharacterSetModificationKind.Add:
+                        if (rule.Characters.Contains(ch))
+                        {
+                            return true;
+                        }
+                        continue;
+
+                    case CharacterSetModificationKind.Remove:
+                        if (rule.Characters.Contains(ch))
+                        {
+                            return false;
+                        }
+                        continue;
+
+                    case CharacterSetModificationKind.Replace:
+                        return rule.Characters.Contains(ch);
+                }
+            }
+
+            // general rule: if the filtering text exactly matches the start of the item then it must be a filter character
+            if (TextTypedSoFarMatchesItem(item, textTypedSoFar))
+            {
+                return false;
+            }
+
+            // Fall back to the default rules for this language's completion service.
+            return completionRules.DefaultCommitCharacters.IndexOf(ch) >= 0;
+        }
+
+        internal static bool TextTypedSoFarMatchesItem(CompletionItem item, string textTypedSoFar)
+        {
+            if (textTypedSoFar.Length > 0)
+            {
+                return item.DisplayText.StartsWith(textTypedSoFar, StringComparison.CurrentCultureIgnoreCase) ||
+                       item.FilterText.StartsWith(textTypedSoFar, StringComparison.CurrentCultureIgnoreCase);
```

Comment:
Didn't we also do filtering explicitly in en-us for people using non-English things trying to type English framework names?